### PR TITLE
Cleanup hero rendering code as a part of Adventure Map rendering

### PR DESCRIPTION
**Pull request by ihhub**
- fix image saving code
- fix boat flag asset issue: the flag suppose to be 4 pixel height but in reality 248
- update comments in relation to Adventure Map rendering
- huge cleanup of hero rendering
- add the code to split hero sprites into tiles needed for the upcoming Adventure Map rendering

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @Branikolog and @LeHerosInconnu, I need your quick help of verifying that hero rendering remains the same as before with this pull request. I don't want to break rendering just now :)

**Review by oleg-derevenetz — CHANGES_REQUESTED** [review]
Hi @ihhub I just left one suggestion, please take a look when you have time.

File: src/engine/image_tool.cpp [hunk 1]
```diff
@@ -96,10 +99,16 @@ namespace
...
 #endif
 
-        const uint32_t width = image.width();
-        const uint32_t height = image.height();
+        if ( surface->pitch != width ) {
+            const uint8_t * imageIn = image.image();
 
-        memcpy( surface->pixels, image.image(), width * height );
+            for ( int32_t i = 0; i < height; ++i ) {
+                memcpy( reinterpret_cast<int8_t *>( surface->pixels ) + surface->pitch * i, imageIn + width * i, static_cast<size_t>( width ) );
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/engine/image_tool.cpp:106]: `int8_t` is a copy-paste mistake from screen.cpp. I will fix there as well.

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> Hi @Branikolog and @LeHerosInconnu, I need your quick help of verifying that hero rendering remains the same as before with this pull request. I don't want to break rendering just now :)

I did a test up to Month: 2 Week: 3 Day: 5 on the Broken Alliance scenario, :) everything seems fine.
Should I do a longer test?

Another thing.
For a while now, when I download files to test PRs, the downloaded file is put in an additional .zip file.
For example, the file "fheroes2_windows_x64_SDL2.zip" is saved as "fheroes2_windows_x64_SDL2.zip.zip".
And this is not the case for the files of the Snapshots (latest builds).
Any idea why this happens?

**Comment by oleg-derevenetz** [comment]
@LeHerosInconnu

> For a while now, when I download files to test PRs, the downloaded file is put in an additional .zip file.

That's because GitHub [creates additional ZIP file on the fly](https://github.com/actions/upload-artifact#limitations) when the artifact is downloaded. Currently there is no workaround.

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by ihhub** [comment]
@LeHerosInconnu , thank you!

**Pull request merged**